ctapipe/ml/stereo_combination.py

```diff
@@ -0,0 +1,249 @@
+from abc import abstractmethod
+
+import astropy.units as u
+import numpy as np
+from astropy.table import Table
+
+from ctapipe.core import Component, Container
+from ctapipe.core.traits import CaselessStrEnum, Unicode
+
+from ..containers import (
+    ArrayEventContainer,
+    ParticleClassificationContainer,
+    ReconstructedEnergyContainer,
+)
+
+
+def _calculate_ufunc_of_telescope_values(tel_data, n_array_events, indices, ufunc):
+    combined_values = np.zeros(n_array_events)
+    ufunc.at(combined_values, indices, tel_data)
+    return combined_values
+
+
+def _weighted_mean_ufunc(tel_values, weights, n_array_events, indices):
+    # avoid numerical problems by very large or small weights
+    weights = weights / weights.max()
+    sum_prediction = _calculate_ufunc_of_telescope_values(
+        tel_values * weights,
+        n_array_events,
+        indices,
+        np.add,
+    )
+    sum_of_weights = _calculate_ufunc_of_telescope_values(
+        weights, n_array_events, indices, np.add
+    )
+    mean = np.full(n_array_events, np.nan)
+    valid = sum_of_weights > 0
+    mean[valid] = sum_prediction[valid] / sum_of_weights[valid]
+    return mean
+
+
+class StereoCombiner(Component):
+    # TODO: Add quality query (after #1888)
+    algorithm = Unicode().tag(config=True)
+    combine_property = CaselessStrEnum(["energy", "classification", "direction"]).tag(
+        config=True
+    )
+
+    @abstractmethod
+    def __call__(self, event: ArrayEventContainer) -> None:
+        """
+        Fill event container with stereo predictions
+        """
+
+    @abstractmethod
+    def predict_table(self, mono_predictions: Table) -> Table:
+        """
+        Constructs stereo predictions from a table of
+        telescope events.
+        """
+
+
+class StereoMeanCombiner(StereoCombiner):
+    """
+    Calculates array-wide (stereo) predictions as the mean of
+    the reconstruction on telescope-level with an optional weighting.
+    """
+
+    weights = CaselessStrEnum(
+        ["none", "intensity", "konrad"],
+        default_value="none",
+    ).tag(config=True)
+
+    def _calculate_weights(self, data):
+        """"""
+
+        if isinstance(data, Container):
+            if self.weights == "intensity":
```

Comment:
Would it make sense here to use an ExpressionEngine rather than hard-coded list of weights? That way one could say
```yaml
weights:  "hillas_intensity * hillas_length / hillas_width"
```

Reply:
At the very least, rename the "konrad" option something understandable like "aspect_weighted_intensity" (width/length is the aspect ratio of the ellipse)